tools/multisig/multisig.go

```diff
@@ -0,0 +1,235 @@
+package multisig
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/manifest"
+	filTypes "github.com/filecoin-project/lotus/chain/types"
+	"github.com/zondax/fil-parser/actors"
+	"github.com/zondax/fil-parser/parser"
+	"github.com/zondax/fil-parser/parser/helper"
+	"github.com/zondax/fil-parser/tools"
+	"github.com/zondax/fil-parser/types"
+	"go.uber.org/zap"
+)
+
+const (
+	VERSION       = "V1"
+	MultisigStore = "multisig"
+
+	metadataParams     = "Params"
+	metadataReturn     = "Return"
+	metadataTxnIDField = "TxnID"
+	metadataIDField    = "ID"
+	metadataMethod     = "Method"
+	metadataValue      = "Value"
+
+	txStatusOk = "ok"
+)
+
+var proposeTranslateMap = map[string]string{
+	parser.MethodPropose:         parser.MethodPropose,
+	parser.MethodProposeExported: parser.MethodPropose,
+}
+
+var cancelApproveTranslateMap = map[string]string{
+	parser.MethodApprove:         parser.MethodApprove,
+	parser.MethodApproveExported: parser.MethodApprove,
+	parser.MethodCancel:          parser.MethodCancel,
+	parser.MethodCancelExported:  parser.MethodCancel,
+}
+
+type EventGenerator interface {
+	GenerateMultisigEvents(ctx context.Context, transactions []*types.Transaction, tipsetCid string, tipsetKey filTypes.TipSetKey) (*types.MultisigEvents, error)
+}
+
+type eventGenerator struct {
+	helper *helper.Helper
+	logger *zap.Logger
+}
+
+func NewEventGenerator(helper *helper.Helper, logger *zap.Logger) EventGenerator {
+	return &eventGenerator{
+		helper: helper,
+		logger: logger,
+	}
+}
+
+func (eg *eventGenerator) GenerateMultisigEvents(ctx context.Context, transactions []*types.Transaction, tipsetCid string, tipsetKey filTypes.TipSetKey) (*types.MultisigEvents, error) {
```

Comment:
isn't `tipsetCid` and `tipsetKey` already in `Transaction.TxBasicBlockData` ? If not I think it should be

Reply:
No, that's why I need to pass the tipster key through this method